src/browser/web_request_handler.ts

```diff
@@ -0,0 +1,145 @@
+limitations under the License.
+*/
+
+/*
+    Intercept and modify the contents of a request at various stages of its lifetime, based on
+    https://github.com/openfin/runtime/blob/develop/docs/api/web-request.md
+
+    v1: handler for onBeforeSendHeaders
+ */
+
+const coreState = require('./core_state');
+const electronApp = require('electron').app;
+const {session, webContents} = require('electron');
+
+import * as Shapes from '../shapes';
+
+const moduleName: string = 'WebRequestHandlers';  // for logging
+
+// passed to webRequest.onBeforeSendHeaders
+interface RequestDetails {
+    id: number;
+    url: string;
+    method: string;
+    resourceType: string;
+    requestHeaders: any;
+    renderProcessId?: number; // not set if the request is not associated with a window
+    renderFrameId?: number;
+}
+
+// passed to callback of webRequest.onBeforeSendHeaders
+interface HeadersResponse {
+    cancel: boolean;
+    requestHeaders?: any;
+}
+
+interface AppUuidFilterUrlMap {
+    [uuid: string]: Shapes.WebRequestHeaderConfig[];  // app UUID => custom header settings
+}
+const filterMap: AppUuidFilterUrlMap = {};
+
+interface UrlHeaderMap {
+    [urlPattern: string]: Shapes.WebRequestHeader[];  // URL pattern => list of headers
+}
+let headerMap: UrlHeaderMap = {};  // reset for every app started and closed
+
+function matchUrlPatterns(url: string, config: Shapes.WebRequestHeaderConfig): boolean {
+    let match: boolean = false;
+    if (config.urlPatterns && config.urlPatterns.length > 0) {
+            match = electronApp.matchesURL(url, config.urlPatterns);
+    }
+    return match;
+}
+
+function applyHeaders(requestHeaders: any, config: Shapes.WebRequestHeaderConfig): void {
+    if (config.headers && config.headers.length > 0) {
+        config.headers.forEach((header) => {
+            Object.keys(header).forEach(key => {
+                requestHeaders[key] = header[key];
+                electronApp.vlog(1, `${moduleName}:beforeSendHeadersHandler setting ${key} = ${header[key]}`);
+            });
+        });
+    }
+}
+
+function beforeSendHeadersHandler(details: RequestDetails, callback: (response: HeadersResponse) => void): void {
+    electronApp.vlog(1, `${moduleName}:beforeSendHeadersHandler for ${JSON.stringify(details)}`);
+    let headerAdded: boolean = false;
+
+    if (details.renderProcessId && details.renderFrameId) {
+        const wc = webContents.fromProcessAndFrameIds(details.renderProcessId, details.renderFrameId);
+        if (wc) {
+            electronApp.vlog(1, `${moduleName}:beforeSendHeadersHandler got webcontents ${wc.id}`);
+            const bw = wc.getOwnerBrowserWindow();
+            const opts: Shapes.WindowOptions = coreState.getWindowOptionsById(bw.id);
+            electronApp.vlog(1, `${moduleName}:beforeSendHeadersHandler window opts ${JSON.stringify(opts)}`);
+            if (opts.customRequestHeaders) {
+                for (const rhItem of opts.customRequestHeaders) {
```

Comment:
Instead of `rhItem` I would use the same identifier as for 2nd param to `matchUrlPatterns` function defined above